src/dscanner/analysis/redundant_attributes.d

```diff
@@ -4,174 +4,60 @@
-	mixin ScopedVisit!ConditionalDeclaration;
-
-private:
-
-	alias ConstAttribute = const Attribute;
-	alias CurrentScope = ConstAttribute[];
-	ref CurrentScope currentAttributes()
-	{
-		return stack[$ - 1];
-	}
-
-	CurrentScope[] stack;
-
-	void addAttribute(const Attribute attr)
-	{
-		removeOverwrite(attr);
-		if (checkAttribute(attr))
-		{
-			currentAttributes ~= attr;
-		}
-	}
-
-	bool checkAttribute(const Attribute attr)
-	{
-		auto match = currentAttributes.find!(a => a.attribute.type == attr.attribute.type);
-		if (!match.empty)
-		{
-			auto token = attr.attribute;
-			addErrorMessage(token.line, token.column, KEY,
-					text("same visibility attribute used as defined on line ",
-						match.front.attribute.line.to!string, "."));
-			return false;
-		}
-		return true;
-	}
-
-	void removeOverwrite(const Attribute attr)
-	{
-		import std.array : array;
-		const group = getAttributeGroup(attr);
-		if (currentAttributes.filter!(a => getAttributeGroup(a) == group
-					&& !isIdenticalAttribute(a, attr)).walkLength > 0)
-		{
-			currentAttributes = currentAttributes.filter!(a => getAttributeGroup(a) != group
-					|| isIdenticalAttribute(a, attr)).array;
-		}
-	}
-
-	bool filterAttributes(const Attribute attr)
-	{
-		return isAccessSpecifier(attr);
-	}
-
-	static int getAttributeGroup(const Attribute attr)
-	{
-		if (isAccessSpecifier(attr))
-			return 1;
-
-		// TODO: not implemented
-		return attr.attribute.type;
-	}
-
-	static bool isAccessSpecifier(const Attribute attr)
-	{
-		auto type = attr.attribute.type;
-		return type.among(tok!"private", tok!"protected", tok!"public", tok!"package", tok!"export") > 0;
-	}
-
-	static bool isIdenticalAttribute(const Attribute a, const Attribute b)
-	{
-		return a.attribute.type == b.attribute.type;
-	}
-
-	auto attributesString()
-	{
-		return currentAttributes.map!(a => a.attribute.type.str).joiner(",").to!string;
+		super(fileName);
 	}
 
 	template ScopedVisit(NodeType)
 	{
-		override void visit(const NodeType n)
+		override void visit(NodeType n)
 		{
-			pushScope();
-			n.accept(this);
-			popScope();
+			ASTBase.Visibility.Kind prevVisibility = currVisibility;
+			currVisibility = ASTBase.Visibility.Kind.undefined;
+			super.visit(n);
+			currVisibility = prevVisibility;
 		}
 	}
 
-	void pushScope()
-	{
-		stack.length++;
-	}
+	mixin ScopedVisit!(AST.StructDeclaration);
+	mixin ScopedVisit!(AST.ClassDeclaration);
+	mixin ScopedVisit!(AST.InterfaceDeclaration);
+	mixin ScopedVisit!(AST.UnionDeclaration);
+	mixin ScopedVisit!(AST.StaticIfCondition);
+	mixin ScopedVisit!(AST.StaticIfDeclaration);
+	mixin ScopedVisit!(AST.TemplateDeclaration);
+	mixin ScopedVisit!(AST.ConditionalDeclaration);
 
-	void popScope()
+	override void visit(AST.VisibilityDeclaration vd)
 	{
-		stack.length--;
+		if (currVisibility == vd.visibility.kind)
+		addErrorMessage(cast(ulong) vd.loc.linnum, cast(ulong) vd.loc.charnum, KEY,
+			"Same visibility attribute used as defined on line %u.".format(currLine));
```

Comment:
Why change this to capital letter? It only makes the diff larger. Also, aren't errors generally starting with a small letter in d-scanner?

Reply:
Nope. It's fine, I just want to make sure that we are consistent.